### ISSUE TITLE
Question about input normalization

### ISSUE THREAD

**Issue by yunjey** [mseitzer/pytorch-fid#3]
Hi, 

First of all, thanks for the nice implementation. Could you explain why you normalized inputs like [here](https://github.com/mseitzer/pytorch-fid/blob/master/inception.py#L127-L131)?
The normalized results seems different from below.

```python
# mean and var for ImageNet
mean = (0.485, 0.456, 0.406)
std = (0.229, 0.224, 0.225)

transform = transforms.Compose([
                    transforms.ToTensor(),
                    transforms.Normalize(mean=mean, std=std)])
```

Best,
Yunjey

**Comment by mseitzer** [comment]
Hi,

thank you for spotting this. This is indeed a bug.

I initially copied the normalization code from the Pytorch implementation (https://github.com/pytorch/vision/blob/f87a896f170a34502a60cdc358c19d8b55e72f54/torchvision/models/inception.py#L73-L76), naively assuming that it normalizes the input to the format the pretrained network expects.

But looking into it more, what this really does is transform from inputs that have been normalized using 
```
mean = (0.485, 0.456, 0.406)
std = (0.229, 0.224, 0.225)

transform = transforms.Compose([
                    transforms.ToTensor(),
                    transforms.Normalize(mean=mean, std=std)])
```
to inputs in range [-1, 1], which is what Tensorflow's pretrained Inception network expects. So it reverses the zero mean, unit variance normalization.

I will submit a fix for this shortly. In the meantime, you could scale your images to range [-1, 1] from range [0, 1] using `2 * x - 1` and run with `transform_input=false`.

Best.
Max

**Comment by XavierXiao** [comment]
I still has a question about normalization. As you said, the Tensorflow's pretrained Inception network expects input in [-1,1], but if you load pretrained pytorch models, it sees like it expects inputs that are in [0,1] and then normalized in a specific way. Quote from pytorch official site:

> All pre-trained models expect input images normalized in the same way, i.e. mini-batches of 3-channel RGB images of shape (3 x H x W), where H and W are expected to be at least 224. The images have to be loaded in to a range of [0, 1] and then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225].

So if you are using the pretrained pytorch model, you should normalize the imgae in this way, rather than to [-1,1] right?

**Comment by mseitzer** [comment]
As I said above, Inception expects inputs to be in [-1, 1]. This is true for Pytorch's Inception as well, as they just ported Tensorflow's weights to Pytorch. So what the Pytorch implementation internally does is it transforms the input from mean/std normalized images to [-1, 1]. This is so that all torchvision models take the same input format, namely mean/std normalized images, as you quoted above.

**Comment by AtlantixJJ** [comment]
Hi, I noticed you used [0,1] image scale in the code

```python
        images = np.array([imread(str(f)).astype(np.float32)
                           for f in files[start:end]])

        # Reshape to (n_images, 3, height, width)
        images = images.transpose((0, 3, 1, 2))
        images /= 255
```

while TF's version is using [-1, 1] image scale. Is this a bug?

**Comment by mseitzer** [comment]
Scaling to range [-1, 1] happens inside the model:

https://github.com/mseitzer/pytorch-fid/blob/4e366b2fc9fb933bec9f6f24c5e87c3bd9452eda/inception.py#L131

**Comment by GiangHLe** [comment]
@mseitzer hi, I know it's late, but I'm trying to implement a more flexible FID version based on your code. I knew that you scale the input from [0,1] to [-1,1] but as I know, the Inception model expects the input image is scaled by ImageNet normalized factor which are `mean = [0.485, 0.456, 0.406]` and `std = [0.229, 0.224, 0.225]`. Therefore, it will be better if we modify the transform when we read data. Am I right?

**Comment by SushantGautam** [comment]
> @mseitzer hi, I know it's late, but I'm trying to implement a more flexible FID version based on your code. I knew that you scale the input from [0,1] to [-1,1] but as I know, the Inception model expects the input image is scaled by ImageNet normalized factor which are `mean = [0.485, 0.456, 0.406]` and `std = [0.229, 0.224, 0.225]`. Therefore, it will be better if we modify the transform when we read data. Am I right?

any answer to this?

**Comment by GiangHLe** [comment]
@SushantGautam I reviewed the original TTUR repository, and it appears that the Inception model used in [their code](https://github.com/bioinf-jku/TTUR/blob/73ab375cdf952a12686d9aa7978567771084da42/FIDvsINC/fid.py#L42) is derived from the [Inception score repository](https://github.com/openai/improved-gan/blob/master/inception_score/model.py). You can download the Inception model from the following link: http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz.

Based on the graph analysis, it seems that the model expects an input range of `x ~ [0, 255]` and applies preprocessing with the formula: `x = (x-128) / 128`.
